lib/vnet/vnet.go

```diff
@@ -212,24 +214,45 @@ func installVnetRoutes(stack *stack.Stack) error {
 	return nil
 }
 
-// Run starts the VNet.
+// Run starts the VNet. It blocks until [ctx] is canceled, at which point it closes the link endpoint, waits
+// for all goroutines to terminate, and destroys the networking stack.
 func (m *Manager) Run(ctx context.Context) error {
-	m.slog.With("ipv6_prefix", m.ipv6Prefix).InfoContext(ctx, "Running Teleport VNet.")
+	m.slog.InfoContext(ctx, "Running Teleport VNet.", "ipv6_prefix", m.ipv6Prefix)
+
+	ctx, cancel := context.WithCancel(ctx)
+
+	allErrors := make(chan error, 3)
 	g, ctx := errgroup.WithContext(ctx)
-	g.Go(func() error { return m.statsHandler(ctx) })
 	g.Go(func() error {
-		return forwardBetweenTunAndNetstack(ctx, m.tun, m.linkEndpoint)
+		defer cancel()
+		err := trace.Wrap(m.statsHandler(ctx))
+		allErrors <- err
+		return err
+	})
+	g.Go(func() error {
+		defer cancel()
+		err := forwardBetweenTunAndNetstack(ctx, m.tun, m.linkEndpoint)
+		allErrors <- err
+		return err
+	})
+	g.Go(func() error {
+		// When the context is canceled for any reason, the caller may have canceled it or one of the other
+		// concurrent tasks, destroy everything and quit.
+		<-ctx.Done()
+		close(m.destroyed)
+		m.linkEndpoint.Close()
+		err := trace.Wrap(m.tun.Close(), "closing TUN device")
+		allErrors <- err
+		return err
 	})
-	return trace.Wrap(g.Wait())
-}
 
-// Destroy closes the link endpoint, waits for all goroutines to terminate, and destroys the networking stack.
-func (m *Manager) Destroy() error {
-	close(m.destroyed)
-	m.linkEndpoint.Close()
+	// Deliberately ignoring the error from g.Wait() to return an aggregate of all errors.
+	_ = g.Wait()
 	m.wg.Wait()
 	m.stack.Destroy()
-	return nil
+
+	close(allErrors)
+	return trace.NewAggregateFromChannel(allErrors, context.Background())
```

Comment:
Consider adding something like these comments / rearrangement to try and keep the order/reasons for all of the various concurrency mechanisms easy to recall going forward (note that I may have somethings wrong here, this is just what I inferred based on studying the code):

```go
	g.Go(func() error {
		// When the context is canceled for any reason (the caller may have canceled it or one of the other
		// concurrent tasks), ensure the `forwardBetweenTunAndNetstack` goroutine is stopped. (The `statsHandler`)
		// goroutine will stop on its own when the context is canceled.)
		<-ctx.Done()

		// Close the link endpoint and TUN device to cause the `forwardBetweenTunAndNetstack` function to
		// return.
		m.linkEndpoint.Close()
		err := trace.Wrap(m.tun.Close(), "closing TUN device")

		allErrors <- err

		return err
	})
	// Deliberately ignoring the error from g.Wait() to return an aggregate of all errors.
	_ = g.Wait()
	// Close the destroyed channel to signal that the VNet is no longer running, and that any in
	// flight connections in the `stack` should be closed.
	close(m.destroyed)
	// Wait for all in flight connections to be closed.
	m.wg.Wait()
	// Now we can destroy the `stack` to free all resources.
	m.stack.Destroy()

	close(allErrors)
	return trace.NewAggregateFromChannel(allErrors, context.Background())
```

Reply:
thanks for the review! I added a bunch of comments in [ebef36a](https://github.com/gravitational/teleport/pull/40893/commits/ebef36afdf3c9b12b562fc7e74fc1109c8a1ac1a)